lib/middleware-common/src/metering.rs

```diff
@@ -0,0 +1,90 @@
+use wasmer_runtime_core::{
+    codegen::{Event, EventSink, FunctionMiddleware, InternalEvent},
+    module::ModuleInfo,
+    wasmparser::{Operator, Type as WpType},
+};
+
+pub struct Metering {
+    limit: u64,
+    current_block: u64,
+}
+
+impl Metering {
+    pub fn new(limit: u64) -> Metering {
+        Metering {
+            limit,
+            current_block: 0,
+        }
+    }
+}
+
+impl FunctionMiddleware for Metering {
+    type Error = String;
+    fn feed_event<'a, 'b: 'a>(
+        &mut self,
+        op: Event<'a, 'b>,
+        _module_info: &ModuleInfo,
+        sink: &mut EventSink<'a, 'b>,
+    ) -> Result<(), Self::Error> {
+        match op {
+            Event::Internal(InternalEvent::FunctionBegin(_)) => {
+                self.current_block = 0;
+            }
+            Event::Wasm(&ref op) | Event::WasmOwned(ref op) => {
+                self.current_block += 1;
+                match *op {
+                    Operator::Loop { .. }
+                    | Operator::Block { .. }
+                    | Operator::End
+                    | Operator::If { .. }
+                    | Operator::Else
+                    | Operator::Unreachable
+                    | Operator::Br { .. }
+                    | Operator::BrTable { .. }
+                    | Operator::BrIf { .. }
+                    | Operator::Call { .. }
+                    | Operator::CallIndirect { .. }
+                    | Operator::Return => {
+                        sink.push(Event::Internal(InternalEvent::GetInternal(0)));
```

Comment:
Maybe we could name the internals indexes using const or enum?
```
enum Internals {
  Points = 0
}
```

Reply:
Changed to dynamically allocating indexes.